File: leetcode/79-word-search.py
# https://leetcode.com/problems/word-search/ (top-interview-questions)

from typing import List


class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        ROWS, COLS = len(board), len(board[0])
        if len(word) > (ROWS*COLS): return False
        visited = set()


        def dfs(r, c, i):
            if i == len(word): return True
            if (r < 0 or c < 0 or r >= ROWS or c >= COLS or word[i] != board[r][c] or (r, c) in visited):
                return False
            visited.add((r, c))
            result = (dfs(r+1, c, i+1) or dfs(r-1, c, i+1) or dfs(r, c+1, i+1) or dfs(r, c-1, i+1))
            visited.remove((r, c))
            return result

        for row in range(ROWS):
            for col in range(COLS):
                if dfs(row, col, 0): return True
        
        return False

File: leetcode/79-word-search.js
// https://leetcode.com/problems/word-search/ (top-interview-questions)

/**
 * @param {character[][]} board
 * @param {string} word
 * @return {boolean}
 */
var exist = function(board, word) {
  const ROWS = board.length;
  const COLS = board[0].length;
  // edge case
  if (word.length > ROWS*COLS) return false;

  const visited = new Set();
  const backtrack = (r, c, i) => {
    if (i === word.length) {
      return true;
    }

    const key = r + ',' + c;

    if (r < 0 || c < 0 || r >= ROWS || c >= COLS || word[i] !== board[r][c] || visited.has(key)) {
      return false;
    }

    visited.add(key);
    const result = backtrack(r+1, c, i+1) || backtrack(r, c+1, i+1) || backtrack(r-1, c, i+1) || backtrack(r, c-1, i+1);
    visited.delete(key);
    return result;
  }

  for (let r = 0; r < ROWS; r++) {
    for (let c = 0; c < COLS; c++) {
      if (backtrack(r, c, 0)) {
        return true;
      }
    }
  }

  return false;
};
